File: app/src/main/java/com/example/simonastojanovic/memeapp/domain/FetchMemeUseCaseImpl.java
package com.example.simonastojanovic.memeapp.domain;

import android.arch.lifecycle.MutableLiveData;

import com.example.simonastojanovic.memeapp.model.Data;
import com.example.simonastojanovic.memeapp.model.MemeRepository;
import com.example.simonastojanovic.memeapp.model.MemesItem;
import com.example.simonastojanovic.memeapp.model.Response;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class FetchMemeUseCaseImpl implements FetchMemeUseCase {

    private MemeRepository memeRepository = new MemeRepository();

    @Override
    public void execute(final MutableLiveData<ArrayList<MemesItem>> memeLiveData) {
        memeRepository.getMemeList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Response>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(Response response) {
                        Data memes = response.getData();

                        List<MemesItem> memeList = memes.getMemes();

                        ArrayList<MemesItem> memeArrayList = new ArrayList<>();

                        for (int i = 0; i < memeList.size(); i++) {
                            MemesItem memeItem = new MemesItem();
                            memeItem.setUrl(memeList.get(i).getUrl());
                            memeArrayList.add(memeItem);
                        }

                        memeLiveData.setValue(memeArrayList);

                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/com/example/simonastojanovic/memeapp/domain/FetchMemeUseCaseImpl.kt
package com.example.simonastojanovic.memeapp.domain

import android.arch.lifecycle.MutableLiveData

import com.example.simonastojanovic.memeapp.model.Data
import com.example.simonastojanovic.memeapp.model.MemeRepository
import com.example.simonastojanovic.memeapp.model.MemesItem
import com.example.simonastojanovic.memeapp.model.Response

import java.util.ArrayList

import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

class FetchMemeUseCaseImpl : FetchMemeUseCase {

    private val memeRepository = MemeRepository()

    override fun execute(memeResponseLiveData: MutableLiveData<ArrayList<MemesItem>>) {
        memeRepository.memeList
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<Response> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onNext(response: Response) {
                        val memes = response.data

                        val memeList = memes!!.memes

                        val memeArrayList = ArrayList<MemesItem>()

                        for (i in memeList.indices) {
                            val memeItem = MemesItem()
                            memeItem.url = memeList[i].url
                            memeArrayList.add(memeItem)
                        }

                        memeResponseLiveData.value = memeArrayList

                    }

                    override fun onError(e: Throwable) {

                    }

                    override fun onComplete() {

                    }
                })
    }
}
